File: feature/standard/oauth/js/OauthFeature.js
const { Oauth2 } = require("./Oauth")

class OauthFeature {
  client
  config
  options

  // Create a new instance for each client instance
  constructor(client, config) {
    this.client = client
    this.config = config
    console.log('OAUTH CONFIG', config)

    this.options = client.options.oauth
    console.log('OAUTH OPTIONS', this.options)
  }

  // If defined, pass in request spec
  async modifyRequest(ctx) {
    const defaultTokenUri = process.env.OAUTH_TOKEN_URI
    const defaultHeaders = {
      accept: 'application/json',
      'Content-Type': 'application/x-www-form-urlencoded',
    }

    const clientSpec = this.options.clientSpec || {
      uri: defaultTokenUri,
      headers: defaultHeaders,
    }

    const refreshSpec = this.options.refreshSpec || {
      uri: defaultTokenUri,
      headers: defaultHeaders,
    }


    const oAuthClient = new Oauth2({
      clientSpec: clientSpec,
      refreshSpec: refreshSpec
    })

    const token = await oAuthClient.getToken()

    ctx.spec.headers.authorization = "Bearer " + token
    return ctx.spec
  }


  // If defined, pass in request spec and response
  // ctx contains context = { start: <time>, end: <time>, res: <response> }
  modifyResult(ctx) {
    return ctx.result
  }

}



module.exports = {
  OauthFeature
}

File: feature/standard/oauth/go/oauth_feature.go
package oauth

import (
	"os"

	"github.com/joho/godotenv"
)

type Options struct {
	ClientSpec  Spec
	RefreshSpec Spec
}

type OauthFeature struct {
	Options Options
}

func (o OauthFeature) modifyRequest(ctx map[string]any) (map[string]any, error) {
	godotenv.Load()

	defaultTokenUri, defaultHeaders := os.Getenv("OAUTH_TOKEN_URI"),
		map[string]string{
			"accept":       "application/json",
			"Content-Type": "application/x-www-form-urlencoded",
		}

	clientSpec := o.Options.ClientSpec
	if clientSpec.Uri == "" {
		clientSpec = Spec{
			Uri:     defaultTokenUri,
			Headers: defaultHeaders,
		}
	}

	refreshSpec := o.Options.RefreshSpec
	if refreshSpec.Uri == "" {
		refreshSpec = Spec{
			Uri:     defaultTokenUri,
			Headers: defaultHeaders,
		}
	}

	oauthClient := &oauth2{
		clientSpec:  clientSpec,
		refreshSpec: refreshSpec,
	}

	token, err := oauthClient.getToken()
	if err != nil {
		return nil, err
	}

	ctx["spec"] = token

	return ctx, nil
}

func (o OauthFeature) modifyResult(ctx map[string]any) (map[string]any, error) {
	return ctx, nil

}

func NewOauthFeature(options Options) *OauthFeature {
	return &OauthFeature{
		Options: options,
	}
}
